### PR TITLE
feat(client): add optional metricsContext metadata

### PR DESCRIPTION
**Pull request by philbooth**
This change is to enable https://github.com/mozilla/fxa-content-server/issues/3551, propagation of metrics context metadata to the auth server for [FxA-70](https://github.com/mozilla/fxa/tree/master/features/FxA-70-kpi-dashboards).

~~Something I wasn't sure about: I call `Object.keys` and `Array.forEach` at one point, are they safe to use in this repo or does it need to be ES3-safe? Happy to replace them with an old-fashioned loop if necessary.~~

**EDIT:** This PR is dependent on https://github.com/mozilla/fxa-auth-server/pull/1213, it can't be merged until that one is.

### CODE REVIEW

File: client/FxAccountClient.js [hunk 1]
```diff
@@ -1031,4 +1099,28 @@ define([
...
+  utmContent: true,
+  utmMedium: true,
+  utmSource: true,
+  utmTerm: true
+};
+
+function marshallMetricsContext (data) {
+  var metricsContext = {};
+
+  Object.keys(data).forEach(function (key) {
```
Comment (philbooth) [inline client/FxAccountClient.js]: ES5 usage here. Is this cool?

**Comment by philbooth** [comment]
@vladikoff, @vbudhram, @shane-tomlinson, r?

File: tests/lib/account.js [hunk 2]
```diff
@@ -164,6 +164,46 @@ define([
           );
       });
 
+      test('#reset password with metricsContext metadata', function () {
+        var user = 'test5' + new Date().getTime();
+        var email = user + '@restmail.net';
+        var passwordForgotToken;
+
+        return respond(client.signUp(email, 'oldpassword'), RequestMocks.signUp)
```
Comment (vbudhram) [inline tests/lib/account.js]: You could potentially use the `accountHelper.newVerifiedAccount` function to save on the boilerplate?

**Comment by vbudhram** [comment]
Besides minor nit, LGTM

File: client/FxAccountClient.js [hunk 3]
```diff
@@ -80,6 +80,18 @@ define([
...
+   *     @param {Number} [flowBeginTime] flow.begin event time
+   *     @param {String} [context] context identifier
+   *     @param {String} [entrypoint] entrypoint identifier
+   *     @param {String} [migration] migration identifier
+   *     @param {String} [service] service identifier
+   *     @param {String} [utm_campaign] marketing campaign identifier
+   *     @param {String} [utm_content] marketing campaign content identifier
+   *     @param {String} [utm_medium] marketing campaign medium
+   *     @param {String} [utm_source] marketing campaign source
+   *     @param {String} [utm_term] marketing campaign search term
```
Comment (vladikoff) [inline client/FxAccountClient.js]: this is not rendering correctly in the YUIDoc: 



Use `grunt doc` to check it out

Comment (vladikoff) [inline client/FxAccountClient.js]: I'm not sure if there is a way to make proper besides doing something like `options.metricsContext.flowId`, need to check the docs.

File: client/FxAccountClient.js (hunk 1 above)
Comment (vladikoff) [inline client/FxAccountClient.js]: If we are planning to get things working in es5 in here: https://github.com/mozilla/fxa-content-server/issues/2886 it might be worth keeping this <es5. Need to discuss...

**Comment by vladikoff** [comment]
If you `npm start` fxa-auth-server and run `npm run test-local` I get test failures: 

```
PASS: #with new device (19ms)
FAIL: #with metricsContext metadata (16ms)
AssertionError: expected { Object (code, errno, ...) } to be falsy
    at assert.notOk (/Users/vladikoff/mozilla/fxa-js-client/node_modules/intern-geezer/assert.js:529:10)
    at onSettled (/Users/vladikoff/mozilla/fxa-js-client/components/p/p.js:357:9)
    at Immediate.flush [as _onImmediate] (/Users/vladikoff/mozilla/fxa-js-client/components/p/p.js:64:6)
    at processImmediate [as _immediateCallback] (timers.js:383:17)
9/10 tests passed
groupEnd: signUp
group: signIn
PASS: #basic (33ms)
PASS: #with keys (45ms)
PASS: #with service (29ms)
PASS: #with reason (39ms)
Waiting for email...
PASS: #incorrect email case (127ms)
Waiting for email...
PASS: #incorrect email case with skipCaseError (111ms)
Waiting for email...
PASS: #incorrectPassword (109ms)
PASS: #with new device (35ms)
PASS: #with existing device (35ms)
FAIL: #with metricsContext metadata (28ms)
AssertionError: expected { Object (code, errno, ...) } to be falsy
    at assert.notOk (/Users/vladikoff/mozilla/fxa-js-client/node_modules/intern-geezer/assert.js:529:10)
    at onSettled (/Users/vladikoff/mozilla/fxa-js-client/components/p/p.js:357:9)
    at Immediate.flush [as _onImmediate] (/Users/vladikoff/mozilla/fxa-js-client/components/p/p.js:64:6)
    at processImmediate [as _immediateCallback] (timers.js:383:17)
9/10 tests passed
groupEnd: signIn
group: verifyCode
Waiting for email...
PASS: #verifyEmail (86ms)
Waiting for email...
PASS: #verifyEmailCheckStatus (67ms)
2/2 tests passed
groupEnd: verifyCode
group: request module
PASS: #heartbeat (4ms)
PASS: #error (2ms)
PASS: #timeout (203ms)
PASS: #bad response format error (56ms)
PASS: #ensure is usable (4ms)
5/5 tests passed
groupEnd: request module
group: hkdf
PASS: #vector 1 (0ms)
PASS: #vector 2 (0ms)
PASS: #vector 3 (1ms)
PASS: #vector 4 (1ms)
4/4 tests passed
groupEnd: hkdf
group: unbundle
PASS: #vector 1 (2ms)
PASS: #vector 2 (2ms)
2/2 tests passed
groupEnd: unbundle
group: credentials
PASS: #client stretch-KDF vectors (17ms)
PASS: #wrap (1ms)
2/2 tests passed
groupEnd: credentials
group: hawkCredentials
PASS: #client derive hawk credentials (1ms)
1/1 tests passed
groupEnd: hawkCredentials
group: passwordChange
Waiting for email...
PASS: #basic (177ms)
Waiting for email...
PASS: #with incorrect case (175ms)
Waiting for email...
PASS: #with incorrect case with skipCaseError (110ms)
Waiting for email...
PASS: #changeFailure (133ms)
4/4 tests passed
groupEnd: passwordChange
group: session
Waiting for email...
PASS: #destroy (98ms)
Waiting for email...
PASS: #status (98ms)
Waiting for email...
PASS: #status error with a false token (99ms)
3/3 tests passed
groupEnd: session
group: certificateSign
Waiting for email...
PASS: #basic (104ms)
Waiting for email...
FAIL: #with metricsContext metadata (104ms)
AssertionError: expected { Object (code, errno, ...) } to be falsy
    at assert.notOk (/Users/vladikoff/mozilla/fxa-js-client/node_modules/intern-geezer/assert.js:529:10)
    at onSettled (/Users/vladikoff/mozilla/fxa-js-client/components/p/p.js:357:9)
    at Immediate.flush [as _onImmediate] (/Users/vladikoff/mozilla/fxa-js-client/components/p/p.js:64:6)
    at processImmediate [as _immediateCallback] (timers.js:383:17)
1/2 tests passed
groupEnd: certificateSign
group: account
Waiting for email...
PASS: #destroy (120ms)
Waiting for email...
PASS: #keys (101ms)
Waiting for email...
PASS: #destroy with incorrect case (142ms)
Waiting for email...
PASS: #destroy with incorrect case with skipCaseError (112ms)
Waiting for email...
PASS: #reset password (64ms)
Waiting for email...
FAIL: #reset password with metricsContext metadata (67ms)
AssertionError: expected { Object (code, errno, ...) } to be falsy
    at assert.notOk (/Users/vladikoff/mozilla/fxa-js-client/node_modules/intern-geezer/assert.js:529:10)
    at onSettled (/Users/vladikoff/mozilla/fxa-js-client/components/p/p.js:357:9)
    at Immediate.flush [as _onImmediate] (/Users/vladikoff/mozilla/fxa-js-client/components/p/p.js:64:6)
    at processImmediate [as _immediateCallback] (timers.js:383:17)
Waiting for email...
```

**Comment by philbooth** [comment]
> I get test failures...

Sorry @vladikoff, I should have mentioned in the PR description: This is dependent on the changes in https://github.com/mozilla/fxa-auth-server/pull/1213/files. So it can't be merged until that one is.

I'll take this out of the `review` column, I was a bit brain-dead when I opened it up yesterday.

**Comment by philbooth** [comment]
Thanks for your other feedback @vladikoff and @vbudhram, fwiw I've pushed fixes for all three.

File: client/FxAccountClient.js [hunk 4]
```diff
@@ -1031,4 +1099,30 @@ define([
...
+  utmContent: true,
+  utmMedium: true,
+  utmSource: true,
+  utmTerm: true
+};
+
+function marshallMetricsContext (data) {
+  var metricsContext = {
+    flowId: data.flowId,
+    flowBeginTime: data.flowBeginTime
```
Comment (vbudhram) [inline client/FxAccountClient.js]: Would it make sense to do any validation on options being set?

Comment (philbooth) [inline client/FxAccountClient.js]: Hmm, I don't *think* so. The auth server validates its payload so doing it here too feels redundant and would open up the can of worms of having the same thing defined in two separate place.

**Comment by philbooth** [comment]
@shane-tomlinson, @vladikoff, now that https://github.com/mozilla/fxa-auth-server/pull/1213 has landed, this one is ready for review again. Sorry about the false alarm last week!

File: client/FxAccountClient.js [hunk 5]
```diff
@@ -1031,4 +1099,30 @@ define([
   return FxAccountClient;
 });
 
+var METRICS_CONTEXT_OPTIONS = {
+  context: true,
+  entrypoint: true,
+  migration: true,
+  service: true,
+  utmCampaign: true,
```
Comment (shane-tomlinson) [inline client/FxAccountClient.js]: Over in the content server, [these fields come from the query parameters](https://github.com/mozilla/fxa-content-server/pull/3587/files#diff-3a32567cd739e05b2dc5a399f75ab73fR15) and are in snake case. I assume the auth server will convert case for heka output?

File: client/FxAccountClient.js [hunk 6]
```diff
@@ -1031,4 +1099,30 @@ define([
...
+  utmContent: true,
+  utmMedium: true,
+  utmSource: true,
+  utmTerm: true
+};
+
+function marshallMetricsContext (data) {
```
Comment (shane-tomlinson) [inline client/FxAccountClient.js]: marshall is an underused word.

**Comment by shane-tomlinson** [comment]
@philbooth - over in [the fxa-70 spec](https://github.com/mozilla/fxa/tree/master/features/FxA-70-kpi-dashboards), `account.deleted` is specified. Do the same parameters need to be passed to `accountDestroy`? The same question applies to `account.verified` and `verifyCode`.

File: client/FxAccountClient.js (hunk 5 above)
Comment (shane-tomlinson) [inline client/FxAccountClient.js]: [yes, the case is converted](https://github.com/mozilla/fxa-auth-server/pull/1213/files#diff-3833ef223ad2c72e2c7db2440f69353eR36).

File: tests/lib/certificateSign.js [hunk 7]
```diff
@@ -44,6 +44,34 @@ define([
...
+              e: '65537'
+            };
+            var duration = 86400000;
+
+            return respond(
+              client.certificateSign(account.signIn.sessionToken, publicKey, duration, {
+                metricsContext: {
+                  flowId: '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef',
+                  flowBeginTime: Date.now(),
+                  forbiddenProperty: 666
```
Comment (shane-tomlinson) [inline tests/lib/certificateSign.js]: My lone suggestion for this PR is to add at least one test that ensures all expected properties make it through, though I remember things being strange with this repo. Is that possible?

**Comment by philbooth** [comment]
> over in the fxa-70 spec, `account.deleted` is specified. Do the same parameters need to be passed to `accountDestroy`? The same question applies to `account.verified` and `verifyCode`.

Those are new activity events and we haven't added them to the auth server yet, or updated those Joi schemata to accept `metricsContext`. The changes here are just to support the properties that were added for existing activity events in https://github.com/mozilla/fxa-content-server/issues/3587.

In hindsight, perhaps it would have been better for me to do all of the activity events in one go and then I could update everything in one go here. But also, I was keen to get something done end-to-end sooner rather than later, so that we can see things working ASAP. Updating everything for the new activity events should be a much quicker process when it can follow the precedents being set for the existing ones.

File: client/FxAccountClient.js (hunk 6 above)
Comment (philbooth) [inline client/FxAccountClient.js]: Ha, I used to be a COM guy once upon a time, so I think that's why it came to me.

**Comment by shane-tomlinson** [comment]
If my test comment is possible, do it, if not, ship it.

File: tests/lib/certificateSign.js (hunk 7 above)
Comment (philbooth) [inline tests/lib/certificateSign.js]: @rfk made a similar point about the auth server tests, which led to the creation of https://github.com/mozilla/fxa-auth-server/issues/1216. The problem is that there are no observable side-effects of this metadata unless we snoop on the log output, which seems tricky. Any bright ideas?

Comment (philbooth) [inline tests/lib/certificateSign.js]: Fwiw, I've pushed some variation in the tests, so that they're at least exercising some different variations of the `metricsContext` properties. Hopefully that's a bit better than what we had before.

Comment (shane-tomlinson) [inline tests/lib/certificateSign.js]: I don't have any, @vladikoff?

**Comment by vladikoff** [comment]
running `npm run test-local` gets me stuck on fxa-auth-server 515b0a3e2ac3159a99b0cf23476dd4f1bc7388cb:
```
Waiting for email...
PASS: #destroy with incorrect case with skipCaseError (99ms)
Waiting for email...
PASS: #reset password (45ms)
Waiting for email...
Waiting for email...
Waiting for email...
Waiting for email...
Waiting for email...
Waiting for email...
```

Tests pass when I checkout fxa-js-client master

**Comment by shane-tomlinson** [comment]
> npm run test-local

Ahha, I forgot about this command. Should we have that run as part of npm test by default?

**Comment by vladikoff** [comment]
> Should we have that run as part of npm test by default? 

what will require the auth server running, we can make that happen

**Comment by philbooth** [comment]
@vladikoff, @shane-tomlinson, have pushed a fix for that test failure, it was down to me trying to be "clever" in the new test case.

File: tests/lib/certificateSign.js (hunk 7 above)
Comment (vladikoff) [inline tests/lib/certificateSign.js]: Yeah we can do deep inspection as part of another PR

File: client/FxAccountClient.js [hunk 8]
```diff
@@ -1031,4 +1099,30 @@ define([
   return FxAccountClient;
 });
 
+var METRICS_CONTEXT_OPTIONS = {
```
Comment (vladikoff) [inline client/FxAccountClient.js]: nit: all the new stuff here, could it be in its own file `metrics.js` or `metrics-contexts.js` or something? 
With the goal of keeping this file cleaner.  Up to you @philbooth @shane-tomlinson

File: client/FxAccountClient.js [hunk 9]
```diff
@@ -1031,4 +1099,30 @@ define([
...
+};
+
+function marshallMetricsContext (data) {
+  var metricsContext = {
+    flowId: data.flowId,
+    flowBeginTime: data.flowBeginTime
+  };
+
+  for (var key in data) {
+    if (data.hasOwnProperty(key) && data[key] && METRICS_CONTEXT_OPTIONS[key]) {
```
Comment (vladikoff) [inline client/FxAccountClient.js]: nit: possible to use `.prototype.hasOwnProperty` here instead? 

Ref: https://github.com/mozilla/fxa-content-server/blob/7d8c2c0661c4492895d39cc10ebcbd602f4bf596/app/scripts/lib/session.js#L46

**Comment by vladikoff** [comment]
@shane-tomlinson @philbooth looks good, left 2 nits, but feel free to merge this

**Comment by philbooth** [comment]
Merging with nits fixed.
